### ISSUE TITLE
ALC233 Mic not working

### ISSUE THREAD

**Issue by gustavodlima** [Mirone/AppleHDAPatcher#13]
I've been researching about verbs, codec patching, etc, and I've been trying to patch my ALC233 codec through verbs, but I haven't dumped codec from Linux yet and I tried to discover a better verbs configuration through ALC231 codec data sheet (didn't find the correct one for the model my laptop uses, so I made a little research based on the ALC231 data sheet) that could make my mic working, but without success until now.

I used the ALC233 patch included but mic, as some people pointed out, it's no working. It seems to be a common problem btw.

If you, @Mirone, need some help to fix that, lemme know. Sure I can provide every information you may need.

EDIT: Yep, audio output is working great. It was necessary to use CodecCommander to make it work, as I used to do in Sierra (using High Sierra now), with a verb command to fix output noise using headphones. Input (internal microphone) IS recognised, but there's no input sound at all.

**Comment by Mirone** [comment]
If it's ComboJack(Microphone + headphone or LineIn) try the @KNNSPeed solution, do a search on Insanelymac.com and you will find something about it.

**Comment by gustavodlima** [comment]
Oh, okay. I'll try that one, but do you think that's possible to fix that problem through verbs only? I'm investigating and the fact the device it's already recognised and sometimes it does a noise when you go to SysPref > Sound > Input is making me to believe that can be fixed through it, but I don't have the necessary technical background >yet< to judged the hypothesis. @Mirone

**Comment by KNNSpeed** [comment]
Drivers use verbs. That's just how audio control works. Look at the Linux sources and you'll see it rather clearly. My method just bypasses using Apple's audio implementation almost entirely, exposing to the OS what the official Realtek Linux (and also Windows, for that matter) drivers do but controlling it in a less abstracted way than Apple. The "native Apple way" doesn't actually support all of the functions of certain audio codecs, and what I did was override and supplement existing functionality to enable full, unhindered use and configuration of the audio subsystem. This is what "VerbStub" is for. 

"ComboJack" is merely a set up config and run time script that harnesses the flexibility of VerbStub to properly initialize and control my system's audio jack in the same way as the official Realtek drivers in the Linux kernel. For those with a different codec, it's partly an example of how to port the Linux driver to Mac, and partly an example of how to use functions Mac OS doesn't natively support.

**Comment by gustavodlima** [comment]
@KNNSpeed, it's a pretty interesting solution. I'm curious about why Mirone's patch isn't working anymore just for Internal Mic. I'm gonna compare verbs data included in *hdaconfig.txt* (ALC233 patch folder) with the data generated by codec dump from Linux. I'm a little busy right now, but as soon as possible, I'm gonna try to fix Mirone's patch and do some research about HDA specification, your solution for the problem and I'm definitely going to checkout the Linux sources. Also, I'm gonna test if that solution is gonna work for me too later.

I have to mention that the fact the device is recognised, but still have no input sound detected, it's telling me this verbs data in Mirone's patch may need to be updated now. @Mirone, could you please, gimme some information about how the ALC233 patch was made, so I can double check if something it's actually wrong with the necessary verbs data for the patch to work again on High Sierra? I'd be grateful.

**Comment by Mirone** [comment]
@gustavodlima ,

EN:
I need you to send me a codec_dump,
a codec may have different PinConfigs a very common example
is the ALC269 there are many variants of this codec, well you asked me
if it is possible to make all the modifications in AppleHDA, well in the past I managed to get success with some codecs, but not all work this should
be your case.

PTBR:

Eu preciso que você me envie um codec_dump,
um codec pode ter PinConfigs diferentes um exemplo muito comum
é o de ALC269 há muitas variantes deste codec, bom você me perguntou
se é possivel fazer todas as modificações em AppleHDA, bem no passado eu consegui obter exito com alguns codecs, mas nem todos funcionam esse deve
ser o seu caso.

**Comment by Mirone** [comment]
@KNNSpeed ,

In other words VerbStub works similar to VodooHDA, but it is necessary to make modifications for each codec in question, am I right?

**Comment by gustavodlima** [comment]
@Mirone, I'm a little busy right now, but asaic I'm gonna send you a codec dump. I wonder if this would work, because if PinConfig is the problem, so this problem should've happened on Sierra too, which is not the case. I'll try myself to patch it through the verbs data I'll collect on my codec dump and I'll send to you if I can't get success on it.

**Comment by Mirone** [comment]
@gustavodlima ,
Take your time, bro.

**Comment by gustavodlima** [comment]
Ok, here's what I'm getting from AppleHDAPatcher using the current ALC233 current patch.
_______________________________________________________________________
​Copying kext
Copying zlib files
Adding data entry AFGLowPowerState to plist
Adding string entry Codec to plist
Adding integer entry CodecID to plist
Adding data entry ConfigData to plist
Adding integer entry FuncGroup to plist
Adding integer entry LayoutID to plist
OSX version 10.13.3 detected
Default target codec: 10ec0233 detected.
Patching AppleHDA codec 11d41984 with 10ec0233
3 codec range comparison(s) to patch
Patching range comparison 11d41983
Patching range comparison 10ec0884
Patching range comparison 10ec0261
Unexpected codec range match count: 2 (6 expected)
Aborting with AppleHDA NOT patched
Copying README file
Creating zip file
Done.
_______________________________________________________________________
Heres a Codec Dump:
[codec-dump.txt](https://github.com/Mirone/AppleHDAPatcher/files/1862171/codec-dump.txt)

**Comment by Mirone** [comment]
This error has been fixed.

**Comment by Mirone** [comment]
> I've managed to post a codec dump above. I would be glad if you could take a look at it.

try this kexts:
[AppleHDA-281.51-ALC233_v1.zip](https://github.com/Mirone/AppleHDAPatcher/files/2019085/AppleHDA-281.51-ALC233_v1.zip)
[AppleHDA-281.51-ALC233_v2.zip](https://github.com/Mirone/AppleHDAPatcher/files/2019086/AppleHDA-281.51-ALC233_v2.zip)

**Comment by gustavodlima** [comment]
@Mirone, app error is fixed, sure. I was talking about the issue itself, the problem with ALC233 Mic not working.

I'll try to use these two kexts you provided above. ASAP I'll give a feedback.

**Comment by gustavodlima** [comment]
Well, I know it's been quite some time, but I had to get back here to clarify how I managed to get it to work. Huge thanks to you, @Mirone.

First, I'm now using macOS Mojave. Some problems my setup had are all gone, but how do I managed to get audio working flawlessly with this codec? Well, I had to use a different approach. AppleALC + CodecCommander (to fix the karaoke effect). It should work with High Sierra/Sierra too.

For some reason, putting AppleALC and Lilu just in CLOVER/kexts/Others wasn't enough. AppleALC wasn't loading as pointed by FBPatcher. After putting AppleALC and Lilu into /Library/Extensions, repairing permissions and rebuilding cache, AppleALC was loaded, but no input sound. (manually rebuild cache if using Kext Utility, just in case...)

Searching for alternatives, I found out your patches for ALC3236 into AppleALC. Tested layout-id=4 and guess what, internal microphone started to work! Audio is great now. It's nothing compared with the quality I got with old methods I used to use in Sierra.

**TL;DR** - AppleALC + CodecCommander did the trick. 'layout-id=4' was the right one to get internal microphone working in my particular case.

**Suggestion** - I think you could include those patches you put into AppleALC in the Laptop folder of this project by default or discourage users to use the patcher, as Lilu + plugins are turning into a better alternative for hot patches like this.

**Note:** I wonder if the patch I wrote for it would work with AppleALC too. I'll test it later and check if I got it or not.

Anyway, thanks again.

**Comment by Mirone** [comment]
Thanks for the feedback, but to be honest most of the users are aware that this AppleHDA Patcher is a simplified version of the new version, but that works very well for quick tests and updates, the new AppleHDA Patcher is here:[](https://olarila.com/forum/viewtopic.php?f=81&t=6528)
and a solution for external microphones (comboJack) is here: [](https://olarila.com/forum/viewtopic.php?f=28&t=7782)
